lib/SingleEventManager.lua

```diff
@@ -14,132 +14,145 @@
 	The hooks provided by SingleEventManager pass the associated Roblox object
 	as the first parameter to the callback. This differs from normal
 	Roblox events.
+
+	SingleEventManager's public methods operate in terms of instances and string
+	keys, differentiating between regular events and property changed signals
+	by calling different methods.
+
+	In the internal implementation, everything is handled via indexing by
+	instances and event objects themselves. This allows the code to use the same
+	structures for both kinds of instance event.
 ]]
 
 local SingleEventManager = {}
 
 SingleEventManager.__index = SingleEventManager
 
-local function createHook(rbx, key, method)
+--[[
+	Constructs a `Hook`, which is a bundle containing a method that can be
+	updated, as well as the signal connection.
+]]
+local function createHook(instance, event, method)
 	local hook = {
 		method = method,
-		connection = rbx[key]:Connect(function(...)
-			method(rbx, ...)
-		end)
 	}
 
+	hook.connection = event:Connect(function(...)
+		hook.method(instance, ...)
+	end)
+
 	return hook
 end
 
-local function createChangeHook(rbx, key, method)
-	local hook = {
-		method = method,
-		connection = rbx:GetPropertyChangedSignal(key):Connect(function(...)
-			method(rbx, ...)
-		end)
+function SingleEventManager.new()
+	local self = {
+		-- Map<Instance, Map<Event, Hook>>
+		_hooks = {},
 	}
 
-	return hook
-end
+	setmetatable(self, SingleEventManager)
 
-local function formatChangeKey(key)
-	return ("!PropertyChangeEvent:%s"):format(key)
+	return self
 end
 
-function SingleEventManager.new()
-	local self = {}
+function SingleEventManager:connect(instance, key, method)
+	self:_connectInternal(instance, instance[key], method)
```

Comment:
Do we want to try to check that `key` is pointing to an event here, and throw a nice error?

Reply:
Hm! I think that would be kind of problematic in this case, since we don't have something like `FindFirstChild` to use as a non-throwing variant.